File: linker/lib/linker.rb
class Linker
  attr_reader :reader
  
  def initialize(file_name)
    @reader = Reader.new(file_name)
  end
  
  def link
    clear!
    
    address_parser.parse
    memory_parser.parse
  end
  
  def errors
    @errors ||= {}
  end
  
  def clear!
    [SymbolTable, MemoryMap].each { |clearable| clearable.clear! }
  end
  
  def to_s
    output = SymbolTable.instance.to_s
    output += "\n\n"
    output += MemoryMap.instance.to_s
    output += "\n"
    output
  end
  
  def symbols
    SymbolTable.instance
  end
  
  def memory_map
    MemoryMap.instance
  end
  
private
  
  def address_parser
    @address_parser ||= AddressParser.new(self)
  end
  
  def memory_parser
    @memory_parser ||= MemoryParser.new(self)
  end
  
end

File: java/linker/src/lib/Linker.java
package lib;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Hashtable;

/**
 * User: Ross Kaffenberger
 * Date: Sep 22, 2009
 * Time: 11:33:37 PM
 */
public class Linker {
    private LinkerReader reader;
    private MemoryParser memoryParser = new MemoryParser(this);
    private AddressParser addressParser = new AddressParser(this);

    public Linker(String fileName) throws FileNotFoundException {
        this.reader = new LinkerReader(fileName);
    }

    public void link() throws IOException {
        clear();

        addressParser.parse();
        memoryParser.parse();
    }

    private void clear() {
        SymbolTable.clear();
        MemoryMap.clear();
    }

    public LinkerReader getReader() {
        return reader;
    }

    public String toString() {
        String output = SymbolTable.getInstance().toString();
        output += "\n";
        output += MemoryMap.getInstance().toString();
        output += "\n";
        return output;

    }
}
